include/zephyr/drivers/mm/system_mm.h

```diff
@@ -339,6 +339,48 @@ int sys_mm_drv_update_page_flags(void *virt, uint32_t flags);
 
 int sys_mm_drv_update_region_flags(void *virt, size_t size, uint32_t flags);
 
+/**
+ * @brief Represents an available memory region.
+ *
+ * A memory region that can be used by allocators. Driver defined
+ * attributes can be used to guide the proper usage of each region.
+ */
+struct sys_mm_drv_region {
+	void *addr; /**< Address of the memory region */
+	size_t size; /**< Size of the memory region */
+	uint32_t attr; /**< Driver defined attributes of the memory region */
+};
+
+/* TODO is it safe to assume no valid region has size == 0? */
+/**
+ * @brief Iterates over an array of regions returned by #sys_mm_drv_query_memory_regions
+ */
+#define SYS_MM_DRV_MEMORY_REGION_FOREACH(regions, iter) \
+	for (iter = regions; iter->size; iter++)
+
+/**
+ * @brief Query available memory regions
+ *
+ * Returns an array of available memory regions. One can iterate over
+ * the array using #SYS_MM_DRV_MEMORY_REGION_FOREACH. Note that the last
+ * item of the array is identified by it's size attribute, which is zero.
+ *
+ * @retval regions A possibly empty array of memory regions.
```

Comment:
"An empty array" means null or a 1-element array with first item->size == 0? To be consistent with **SYS_MM_DRV_MEMORY_REGION_FOREACH**

Reply:
Oh, empty means "containing only the sentinel at the end", will make it clearer.